File: src/test/java/fr/marstech/mtlinkspray/service/ShortenerServiceUnitTest.java
package fr.marstech.mtlinkspray.service;

import jakarta.servlet.http.HttpServletRequest;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class ShortenerServiceUnitTest {

    @Test
    void getShortenedLink() {
        // Arrange
        HttpServletRequest mockRequest = mock(HttpServletRequest.class);
        when(mockRequest.getServerName()).thenReturn("localhost");
        when(mockRequest.getServerPort()).thenReturn(8080);
        when(mockRequest.getScheme()).thenReturn("http");
        String uid = "abc123";

        // Act
        String result = ShortenerService.getShortenedLink(mockRequest, uid);

        // Assert
        assertNotNull(result);
        assertTrue(result.contains(uid));
        assertTrue(result.startsWith("http://localhost:8080"));
    }

    @Test
    void getShortenedLink_shouldFilterDefaultPort() {
        // Arrange
        HttpServletRequest mockRequest = mock(HttpServletRequest.class);
        when(mockRequest.getServerName()).thenReturn("my-server-name");
        when(mockRequest.getServerPort()).thenReturn(443);
        when(mockRequest.getScheme()).thenReturn("https");
        String uid = "def605";

        // Act
        String result = ShortenerService.getShortenedLink(mockRequest, uid);

        // Assert
        assertNotNull(result);
        assertTrue(result.contains(uid));
        assertTrue(result.startsWith("https://my-server-name/"));
    }
}

File: src/test/kotlin/fr/marstech/mtlinkspray/service/ShortenerServiceUnitTest.kt
package fr.marstech.mtlinkspray.service

import fr.marstech.mtlinkspray.service.ShortenerService.Companion.getShortenedLink
import jakarta.servlet.http.HttpServletRequest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.mock

internal class ShortenerServiceUnitTest {

    @Test
    fun getShortenedLink() {
        // Arrange
        val mockRequest = mock(HttpServletRequest::class.java)
        Mockito.`when`(mockRequest.serverName).thenReturn("localhost")
        Mockito.`when`(mockRequest.serverPort).thenReturn(8080)
        Mockito.`when`(mockRequest.scheme).thenReturn("http")
        val uid = "abc123"

        // Act
        val result = getShortenedLink(mockRequest, uid)

        // Assert
        Assertions.assertNotNull(result)
        Assertions.assertTrue(result.contains(uid))
        Assertions.assertTrue(result.startsWith("http://localhost:8080"))
    }

    @Test
    fun getShortenedLink_shouldFilterDefaultPort() {
        // Arrange
        val mockRequest = mock(HttpServletRequest::class.java)
        Mockito.`when`(mockRequest.serverName).thenReturn("my-server-name")
        Mockito.`when`(mockRequest.serverPort).thenReturn(443)
        Mockito.`when`(mockRequest.scheme).thenReturn("https")
        val uid = "def605"

        // Act
        val result = getShortenedLink(mockRequest, uid)

        // Assert
        Assertions.assertNotNull(result)
        Assertions.assertTrue(result.contains(uid))
        Assertions.assertTrue(result.startsWith("https://my-server-name/"))
    }

    @Test
    fun getShortenedLink_withBlankUid_returnsNull() {
        val mockRequest = mock(HttpServletRequest::class.java)
        Mockito.`when`(mockRequest.serverName).thenReturn("localhost")
        Mockito.`when`(mockRequest.serverPort).thenReturn(8080)
        Mockito.`when`(mockRequest.scheme).thenReturn("http")
        getShortenedLink(mockRequest, "   ")
        // TODO test exception
    }

    @Test
    fun getShortenedLink_withSpecialCharacters_encodesUid() {
        val mockRequest = mock(HttpServletRequest::class.java)
        Mockito.`when`(mockRequest.serverName).thenReturn("localhost")
        Mockito.`when`(mockRequest.serverPort).thenReturn(8080)
        Mockito.`when`(mockRequest.scheme).thenReturn("http")
        val uid = "abc 123"
        val result = getShortenedLink(mockRequest, uid)
        Assertions.assertTrue(result.contains("abc%20123"))
    }
}
